packages/dds/tree/src/feature-libraries/modular-schema/discrepancies.ts

```diff
@@ -0,0 +1,339 @@
+	| AllowedTypeIncompatibility
+	| FieldKindIncompatibility
+	| ValueSchemaIncompatibility;
+
+type SchemaFactoryNodeKind = "object" | "leaf" | "map";
+
+interface NodeKindIncompatibility {
+	identifier: string;
+	mismatch: "nodeKind";
+	view: SchemaFactoryNodeKind | undefined;
+	stored: SchemaFactoryNodeKind | undefined;
+}
+
+interface NodeFieldsIncompatibility {
+	identifier: string;
+	mismatch: "fields";
+	differences: FieldIncompatibility[];
+}
+
+type NodeIncompatibility = NodeKindIncompatibility | NodeFieldsIncompatibility;
+
+type Incompatibility = FieldIncompatibility | NodeIncompatibility;
+
+/**
+ * The workflow for finding schema incompatibilities:
+ * 1. Identify discrepancies in the root schema.
+ * 2. For each node schema in the `view`:
+ * - Check if the node schema exists in the `stored`.
+ * - If it exists in the `stored`, perform exhaustive validation to identify all discrepancies.
+ * 3. For each node schema in the `stored`, check if it exists in the `view`. The overlapping parts
+ * were already tracked in the previous step.
+ *
+ * @returns the discrepancies between two TreeStoredSchema objects
+ */
+export function getAllowedContentIncompatibilities(
+	view: TreeStoredSchema,
+	stored: TreeStoredSchema,
+): Incompatibility[] {
+	const incompatibilities: Incompatibility[] = [];
+
+	// check root schema discrepancies
+	incompatibilities.push(
+		...trackFieldDiscrepancies(view.rootFieldSchema, stored.rootFieldSchema),
+	);
+
+	// Verify the existence and type of a node schema given its identifier (key), then determine if
+	// an exhaustive search is necessary.
+	const nodeKeySet = new Set<TreeNodeSchemaIdentifier>();
+	for (const [key, viewNodeSchema] of view.nodeSchema) {
+		nodeKeySet.add(key);
+
+		if (viewNodeSchema instanceof ObjectNodeStoredSchema) {
+			if (!stored.nodeSchema.has(key)) {
+				incompatibilities.push({
+					identifier: key,
+					mismatch: "nodeKind",
+					view: "object",
+					stored: undefined,
+				});
+			} else {
+				const storedNodeSchema = stored.nodeSchema.get(key);
+				if (storedNodeSchema instanceof MapNodeStoredSchema) {
+					incompatibilities.push({
+						identifier: key,
+						mismatch: "nodeKind",
+						view: "object",
+						stored: "map",
+					} satisfies NodeKindIncompatibility);
+				} else if (storedNodeSchema instanceof LeafNodeStoredSchema) {
+					incompatibilities.push({
+						identifier: key,
+						mismatch: "nodeKind",
+						view: "object",
+						stored: "leaf",
+					} satisfies NodeKindIncompatibility);
+				} else if (storedNodeSchema instanceof ObjectNodeStoredSchema) {
+					const differences = trackObjectNodeDiscrepancies(viewNodeSchema, storedNodeSchema);
+					if (differences.length > 0) {
+						incompatibilities.push({
+							identifier: key,
+							mismatch: "fields",
+							differences,
+						} satisfies NodeFieldsIncompatibility);
+					}
+				} else {
+					throw new TypeError("This type of Node Stored Schema is not yet supported");
```

Comment:
could you dedupe all of the separate declarations of this into a single file-level helper? Imagining something like this:

```
function throwUnsupportedNodeType(type: string): never {
	throw new TypeError(`Unsupported node stored schema type: ${type}`);
}
```